Create basic_cog.py

Daggy1234 opened the pull request:
## Summary

Adds a new example, that shows a very basic use case for cogs

This includes
- checks
- listeners
- commands
- command groups

## Checklist

<!-- Put an x inside [ ] to check it, like so: [x] -->

- [ ] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [x] This PR is **not** a code change (e.g. documentation, README, ...)

TheOneMusic commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+intents = discord.Intents.default()
+intents.members = True
+
+
```
```suggestion

```
Removing whitespace

TheOneMusic commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+
+
+# We define the basic bot, setting the prefix, description and intents
+
```
```suggestion

```
Removing whitespace

TheOneMusic commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+    A simple utility cog
+    """
+
+    def __init__(self, bot) -> None:
```
```suggestion
    def __init__(self, bot):
```
I don't think type hinting for `__init__` is useful

TheOneMusic commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+
+# We define the basic bot, setting the prefix, description and intents
+
+bot = commands.Bot(command_prefix='?',
+                   description=description, intents=intents)
```
```suggestion
bot = commands.Bot(command_prefix='?', description=description, intents=intents)
```

Rapptz replied:
This was not resolved.

TheOneMusic commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+bot = commands.Bot(command_prefix='?',
+                   description=description, intents=intents)
+
+
```
```suggestion

```

ImVaskel commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+# the class inherits from the commands.Cog class
+# The main advantage is we can take advantage of Object Oriented Programming
+# This also groups all commands in the help command under the `Utility` cog
+class Utility(commands.Cog):
```
```suggestion
class Misc(commands.Cog, name="Miscellaneous"):
```
I think showcasing passing the ability to pass `name` is a good idea.

Daggy1234 replied:
Yup sounds good

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+        # We pass in the bot parameter when initializing the class, and make it a property. This lets us access the bot paramater throughput
+        self.bot = bot
+        # You can also add other properties to the class
+        self.other_vale = "this is a property of the class"
```
```suggestion
        self.other_vale = 'this is a property of the class'
```

mesub7 replied:
```suggestion
        self.other_value = 'This is a property of the class.'
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+        self.other_vale = "this is a property of the class"
+
+    # Cog checks are nice feature in cogs.
+    # They allow you to define a check that works for every single command in the cog
```
```suggestion
    # They allow you to define a check that works for every single command in the cog.
```

mesub7 replied:
```suggestion
    # They allow you to define a check that works for every single command in the cog.
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+
+    # Cog checks are nice feature in cogs.
+    # They allow you to define a check that works for every single command in the cog
+    # You return a boolean value. if the check passes the command runs, otherwise a CheckFailure is raised. 
```
```suggestion
    # You return a boolean value. If the check passes the command runs, otherwise a CheckFailure is raised. 
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+    # They allow you to define a check that works for every single command in the cog
+    # You return a boolean value. if the check passes the command runs, otherwise a CheckFailure is raised. 
+    async def cog_check(self, ctx):
+        # Check if there is a guild the command was invoked in
```
```suggestion
        # Check if there is a guild the command was invoked in.
```

mesub7 replied:
```suggestion
        # Check if there is a guild the command was invoked in.
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+intents.members = True
+
+
+# We define the basic bot, setting the prefix, description and intents
```
```suggestion
# We define the basic bot, setting the prefix, description and intents.
```

mesub7 replied:
```suggestion
# We define the basic bot, setting the prefix, description and intents.
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+                   description=description, intents=intents)
+
+
+# we define the class that will house the Utility commands
```
```suggestion
# We define the class that will house the Utility commands.
```

mesub7 replied:
```suggestion
# We define the class that will house the Utility commands.
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+
+
+# we define the class that will house the Utility commands
+# the class inherits from the commands.Cog class
```
```suggestion
# The class inherits from the `commands.Cog` class.
```

mesub7 replied:
```suggestion
# The class inherits from the `commands.Cog` class.
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+
+# we define the class that will house the Utility commands
+# the class inherits from the commands.Cog class
+# The main advantage is we can take advantage of Object Oriented Programming
```
```suggestion
# The main advantage is we can take advantage of Object Oriented Programming.
```

mesub7 replied:
```suggestion
# The main advantage is we can take advantage of Object Oriented Programming.
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+# we define the class that will house the Utility commands
+# the class inherits from the commands.Cog class
+# The main advantage is we can take advantage of Object Oriented Programming
+# This also groups all commands in the help command under the `Utility` cog
```
```suggestion
# This also groups all commands in the help command under the `Utility` cog.
```

mesub7 replied:
```suggestion
# This also groups all commands in the help command under the `Utility` cog.
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+# This also groups all commands in the help command under the `Utility` cog
+class Utility(commands.Cog):
+    """
+    A simple utility cog
```
```suggestion
    A simple utility cog.
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+
+    def __init__(self, bot) -> None:
+
+        # We pass in the bot parameter when initializing the class, and make it a property. This lets us access the bot paramater throughput
```
```suggestion
        # We pass in the bot parameter when initializing the class, and make it a property. This lets us access the bot paramater throughout.
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+            return True
+        return False
+
+    # A simple basic command inside a cog
```
```suggestion
    # A simple basic command inside a cog.
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+        """Adds two numbers together"""
+        await ctx.send(left + right)
+
+    # A command group inside a cog
```
```suggestion
    # A command group inside a cog.
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+        """Is the bot cool?"""
+        await ctx.send('Yes, the bot is cool.')
+
+    # Listening for events using a cog
```
```suggestion
    # Listening for events using a cog.
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,81 @@
+
+
+# This method allows us to add our Cog to the main bot. Without this you cog won't show up.
+# You can use bot.remove_cog to remove cogs from a bot
```
```suggestion
# You can use bot.remove_cog to remove cogs from a bot.
```

mesub7 reviewed (changes requested):
Looks good but there's some punctuation issues that need addressing.

ImVaskel commented:
I would make an example of `on_cog_command_error` handling the cog check failing.

Daggy1234 commented:
> I would make an example of `on_cog_command_error` handling the cog check failing.

Yeah I get where you're coming from. I do feel  that given there will be a separate example on error handling, including a breif error handling example might not be a good idea. Its best to make this  simple.

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+        await ctx.send('Yes, the bot is cool.')
+
+    # Listening for events using a cog.
+    # This works like the bot.event decorator.
```
A listener is not like an overridden event and this may cause confusion regarding overriden events and behavious that can occur, the `on_message` event for example.
A more apt choice of works would be
`# This is how you create an event listener within a Cog.`

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+            await guild.system_channel.send(to_send)
+
+
+# This method allows us to add our Cog to the main bot. Without this you cog won't show up.
```
```suggestion
# This method allows us to add our Cog to the main bot. Without this the Cog will not be added to `Bot.cogs` and therefore its listeners and commands will not be registered.
```

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+
+
+# This method allows us to add our Cog to the main bot. Without this you cog won't show up.
+# You can use bot.remove_cog to remove cogs from a bot.
```
```suggestion
# You can use bot.remove_cog to remove a Cog from a bot.
```

Trying to stick to one case for 'cog'. As it is a class we should opt for PascalCase.

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+        """Is the bot cool?"""
+        await ctx.send('Yes, the bot is cool.')
+
+    # Listening for events using a cog.
```
```suggestion
    # Listening for events within a Cog.
```

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+        """Adds two numbers together"""
+        await ctx.send(left + right)
+
+    # A command group inside a cog.
```
```suggestion
    # A command group within a Cog.
```

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+        return False
+
+    # A simple basic command inside a cog.
+    # Instead of using bot.command decorator, the commands.command decorator is used.
```
A brief expansion on why we do this, perhaps?

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+            return True
+        return False
+
+    # A simple basic command inside a cog.
```
```suggestion
    # A simple basic command within a Cog.
```

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+    # They allow you to define a check that works for every single command in the cog.
+    # You return a boolean value. If the check passes the command runs, otherwise a `commands.CheckFailure` is raised. 
+    async def cog_check(self, ctx):
+        # Check if there is a guild the command was invoked in.
```
```suggestion
        # Check if the command was invoked within a Guild channel.
```

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+        # You can also add other properties to the class.
+        self.other_value = 'This is a property of the class.'
+
+    # Cog checks are nice feature in cogs.
```
```suggestion
    # Cog checks are nice feature in Cogs.
```

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+        self.other_value = 'This is a property of the class.'
+
+    # Cog checks are nice feature in cogs.
+    # They allow you to define a check that works for every single command in the cog.
```
```suggestion
    # They allow you to define a check that applies to every single command in the Cog.
```

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+
+    # Cog checks are nice feature in cogs.
+    # They allow you to define a check that works for every single command in the cog.
+    # You return a boolean value. If the check passes the command runs, otherwise a `commands.CheckFailure` is raised. 
```
```suggestion
    # You return a boolean value or can raise an Exception. If the check passes the command runs, otherwise a `commands.CheckFailure` is raised. 
```

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+        # We pass in the bot parameter when initializing the class, and make it a property. This lets us access the bot parameter throughout.
+        self.bot = bot
+        # You can also add other properties to the class.
+        self.other_value = 'This is a property of the class.'
```
```suggestion
        self.other_value = 'This is a class attribute.'
```
Not a property here.

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+
+        # We pass in the bot parameter when initializing the class, and make it a property. This lets us access the bot parameter throughout.
+        self.bot = bot
+        # You can also add other properties to the class.
```
```suggestion
        # You can also add other attributes to the class.
```

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+
+    def __init__(self, bot):
+
+        # We pass in the bot parameter when initializing the class, and make it a property. This lets us access the bot parameter throughout.
```
```suggestion
        # We pass in the bot parameter when initializing the class, and make it a class attribute. This lets us access the bot parameter throughout our methods.
```

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+# The class inherits from the `commands.Cog` class.
+# The main advantage is we can take advantage of Object Oriented Programming.
+# This also groups all commands in the help command under the `Misc` cog.
+# The name kwarg passed allows us to set the name of the cog. If not set it defaults to the name of the class.
```
```suggestion
# The name kwarg passed allows us to set the name of the Cog. If not set it will default to the name of the class.
```

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+# We define the class that will house the Miscellaneous commands.
+# The class inherits from the `commands.Cog` class.
+# The main advantage is we can take advantage of Object Oriented Programming.
+# This also groups all commands in the help command under the `Misc` cog.
```
```suggestion
# This also groups all commands in the help command under the `Misc` category.
```

It is named 'category' in the DefaultHelpCommand and is commonly known as this.

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+bot = commands.Bot(command_prefix='?', description=description, intents=intents)
+
+# We define the class that will house the Miscellaneous commands.
+# The class inherits from the `commands.Cog` class.
```
```suggestion
# The class *must* inherit from the `commands.Cog` class.
```

AbstractUmbra commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,78 @@
+The example uses cogs to organize the code
+'''
+
+intents = discord.Intents.default()
```
You should perhaps explain in one small comment why we use the intents here.

AbstractUmbra reviewed (changes requested):
Small nits here and there.

Rapptz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,84 @@
+            return True
+        return False
+
+    # A simple basic command within a Cog.
```
```suggestion
```

Rapptz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,84 @@
+        return False
+
+    # A simple basic command within a Cog.
+    # Instead of using bot.command decorator, the commands.command decorator is used.
+    # This is as the bot.command decorator will add the command to the bot's command list. 
+    # commands.commands.command is used so the command is registered under the Cog's command list.
```
```suggestion
    # Commands within a Cog are defined using the commands.command decorator 
    # instead of the bot.command decorator.
    # They must also pass a `self` parameter since it's within a class.
```

Rapptz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,84 @@
+        """Adds two numbers together"""
+        await ctx.send(left + right)
+
+    # A command group within a Cog.
+    @commands.group()
+    async def cool(ctx):
+        """Says if a user is cool.
+        In reality this just checks if a subcommand is being invoked.
+        """
+        if ctx.invoked_subcommand is None:
+            await ctx.send('No, {0.subcommand_passed} is not cool'.format(ctx))
+
+    @cool.command(name='bot')
+    async def _bot(ctx):
+        """Is the bot cool?"""
+        await ctx.send('Yes, the bot is cool.')
```
Rather than copying from a previous example it's probably better to come up with a newer one.

Rapptz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,84 @@
+        """Is the bot cool?"""
+        await ctx.send('Yes, the bot is cool.')
+
+    # Listening for events within a cog.
+    # This is how you create an event listener for a cog.
```
```suggestion
    # Listening to events within a cog requires the `commands.Cog.listener` decorator.
    # Note that `self` has to be passed for these as well.
```

Rapptz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,84 @@
+            await guild.system_channel.send(to_send)
+
+
+# This method allows us to add our Cog to the main bot. 
+# Without this your  Cog will not be added to `Bot.cogs` and therefore its listeners and commands will not be registered.
+# You can use bot.remove_cog to remove a Cog from a bot.
```
```suggestion
# Cogs have to be explicitly added.
```

Rapptz reviewed (changes requested):
I'm not really a fan of this example because it doesn't really showcase anything that a cog is good at.

The categorisation in this example is basically gone because it went for the "Miscellaneous" route which is a throwaway category so none of the commands are related in any way. The state sharing was not shown at all, in fact none of the commands or listeners even used `self`.

Overall, I think this example needs to be entirely reworked.

Daggy1234 commented:
So I reworked the example,

by trying to show both class methods and the advantages of a shared state in the reworked group. lmk if it's better!

kal-byte commented on examples/basic_cog.py:
```diff
@@ -14,61 +14,80 @@
+    @commands.group(invoke_without_command=True)
+    async def rate(self, ctx):
+        """Let the bot rate you.
+        Generae a random rating between 1 and 100
```
Spelling nit here, should be `generate`.

pikaninja commented on examples/basic_cog.py:
```diff
@@ -77,8 +96,6 @@ async def on_member_join(self, member):
-# This method allows us to add our Cog to the main bot. 
-# Without this your  Cog will not be added to `Bot.cogs` and therefore its listeners and commands will not be registered.
-# You can use bot.remove_cog to remove a Cog from a bot.
+# Cogs have to be explicitly added.
```
Would adding a note about extensions be good here?

Daggy1234 replied:
There would be a separate example for extensions so I feel like I should keep this separate.

pikaninja replied:
well it could just be like "cogs are often used with extensions, you can read about them here: https://discordpy.readthedocs.io/en/latest/ext/commands/extensions.html#ext-commands-extensions which might be good to prompt them to looking at it, but I don't know if that's something that's good for an example.

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,101 @@
+        Generate a random rating between 1 and 100.
+        """
+        # Thie command uses the random stdlib to generate a number
+        # We use teh class method `get_rating`
```
```suggestion
        # We use the class method `get_rating`.
```

Gobot1234 replied:
I would also remove class in "the class method", as it may cause confusion

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,101 @@
+    async def _user(self, ctx, *, user: discord.User = None):
+        """Rate another user"""
+        if not user:
+            return await ctx.send("I need a user to rate.")
```
```suggestion
            return await ctx.send('I need a user to rate.')
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,101 @@
+        if not user:
+            return await ctx.send("I need a user to rate.")
+        rating = self.get_rating(user.name)
+        await ctx.send(f'I have given {user.mention} a rating of `{rating}%`.')
```
```suggestion
        await ctx.send('I have given {} a rating of `{}%`.'.format(user.mention, rating))
```
Keeping .format consistent with other examples in the reop.

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,101 @@
+        rating = self.get_rating(ctx.author.name)
+        await ctx.send(f"{ctx.author.mention} is `{rating}%` cool.")
+
+    # Subcommand for the group
```
```suggestion
    # Subcommand for the group:
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,101 @@
+        # Thie command uses the random stdlib to generate a number
+        # We use teh class method `get_rating`
+        rating = self.get_rating(ctx.author.name)
+        await ctx.send(f"{ctx.author.mention} is `{rating}%` cool.")
```
```suggestion
        await ctx.send('{} is `{}%` cool.'.format(ctx.author.mention, rating))
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,101 @@
+        """Let the bot rate you.
+        Generate a random rating between 1 and 100.
+        """
+        # Thie command uses the random stdlib to generate a number
```
```suggestion
        # This command uses the random stdlib to generate a number.
```

ToxicKidz replied:
```suggestion
        # This command uses the random stdlib to generate a number
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,101 @@
+        # You can also add other attributes to the class.
+        self.ratings = {}
+
+    # Cogs also have the capability to define checks that apply to every command in the Cog
```
```suggestion
    # Cogs also have the capability to define checks that apply to every command in the Cog.
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,101 @@
+                   description=description, intents=intents)
+
+
+# This defines a class that contains commands in a "Miscellaneous" category
+# The library calls this concept a Cog and they must inherit from `commands.Cog`
+# Cogs are useful for grouping commands and having them share state
+# This class will group commands in a "Misc" category in the default help command
+# The `name` keyword argument passed allows us to set the name of the Cog
+# If a name is not given then it will default to the name of the class
```
```suggestion
# This defines a class that contains commands in a "Miscellaneous" category.
# The library calls this concept a Cog and they must inherit from `commands.Cog`.
# Cogs are useful for grouping commands and having them share state.
# This class will group commands in a "Misc" category in the default help command.
# The `name` keyword argument passed allows us to set the name of the Cog.
# If a name is not given then it will default to the name of the class.
```

mesub7 reviewed (changes requested):
One or two typos along with some other nits.

ToxicKidz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,101 @@
+        Generate a random rating between 1 and 100.
+        """
+        # Thie command uses the random stdlib to generate a number
+        # We use teh class method `get_rating`
```
```suggestion
        # We use the class method `get_rating`
```

ToxicKidz reviewed (changes requested):
Spelling mistakes

boehs reviewed (changes requested):
styling

Daggy1234 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,100 @@
+    # They must also pass a `self` parameter since it's within a class.
+    @commands.command()
+    async def add(self, ctx, left: int, right: int):
+        """Adds two numbers together"""
```
Thats a docstring so it shows up in help

Rapptz reviewed (changes requested):
I think there are too many excessive newlines here. 

Formatting aside, I don't like the example used for showcasing state and my concerns with the categorisation were not resolved. It uses member name for some reason which encourages bad practice and it doesn't really make sense for a command to cache ratings anyway.

I'd prefer if the examples were useful.

Daggy1234 commented:
So I rewrote the example away from the rating system and tried out a better example (thank you ilya for the idea). It's a simple afk system where a user can set their afk status. Uses a dictionary to store data.

Rapptz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,116 @@
+        """
+        Set your afk status
+        """
+        u_id = ctx.author.id
+        # Simply check if they exist in our afk dictionary aldready.
+        # If they exist then tell them they are afk, or let them set a new afk status.
+        rating = self.afk.get(u_id)
+        if not rating:
+            self.afk[u_id] = reason
+            await ctx.send(f"{ctx.author.mention} I have set your afk status as `{reason}`.")
+        else:
+            await ctx.send(f"{ctx.author.mdention} you are aldready afk.")
```
```suggestion
        user_id = ctx.author.id
        if user_id not in self.afk_reasons:
            self.afk_reasons[user_id] = reason
            mentions = discord.AllowedMentions.none()
            await ctx.send("I've set your afk status due to {reason}".format(reason), allowed_mentions=mentions)
        else:
            await ctx.send("You are already set as afk.")
```

1. Untrusted user input should be using `AllowedMentions`
2. f-strings can't be used (3.5 minimum support)
3. Too many typos

Rapptz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,116 @@
+    @afk.command(name='user')
+    async def _user(self, ctx, *, user: discord.User = None):
+        """Check a user's afk status"""
+        if not user:
+            return await ctx.send("I need a user to check the afk status")
```
This is definitely not how we do error handling in the library.

Rapptz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,116 @@
+        """Check a user's afk status"""
+        if not user:
+            return await ctx.send("I need a user to check the afk status")
+        afk = self.afk.get(user.id)
+        if afk:
+            return await ctx.send(f"{user.name} is afk with reason `{afk}`")
+        return await ctx.send(f'{user.name} is currently not afk')
```
try-except works better for this case. Also don't use f-strings, use allowed_mentions, etc. just like above.

Rapptz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,116 @@
+        # Don't want to reply to bots
+        if message.author.bot:
+            return
+        # Ensure that the message wasn't the afk command
+        ctx = await self.bot.get_context(message)
+        if ctx.invoked_with == "afk":
+            return
```
This bit of code is wrong. Also reminds me that we really need an event for messages that aren't commands.

```suggestion
        # Ensure that the message wasn't a command
        ctx = await self.bot.get_context(message)
        if ctx.valid:
            return
```

This will have a bit of an odd behaviour (if the afk user uses a command then their afk status won't go away) but you can fix that one.

Rapptz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,116 @@
+            return
+
+        # If an afk user has sent a message, remove their afk
+        m_id = message.author.id
+        afk_status = self.afk.get(m_id)
+        if afk_status:
+            await message.channel.send(f"Welcome back {message.author.mention}! I have removed your afk status.")
+            self.afk.pop(m_id)
+            return
```
1. Use descriptive variable names.
2. Don't use f-strings
3. You can do `self.afk_reasons.pop(member_id, None)` and check for `None`.

Rapptz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,116 @@
+            return
+
+        # If an afk user is mentioned in a message, let the author of the message know they are afk
+        mentions = [user.id for user in message.mentions]
```
```suggestion
        mentions = {user.id: user for user in message.mentions}
```

Dicts have faster lookup than lists. Also this will be used later on for resolving.

Rapptz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,116 @@
+
+        # If an afk user is mentioned in a message, let the author of the message know they are afk
+        mentions = [user.id for user in message.mentions]
+        common = [afk_user for afk_user in self.afk.keys()
+                  if afk_user in mentions]
```
```suggestion
        mentioned_afk_users = [user_id for user_id in self.afk_reasons.keys() if user_id in mentions]
```

Rapptz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,116 @@
+        mentions = [user.id for user in message.mentions]
+        common = [afk_user for afk_user in self.afk.keys()
+                  if afk_user in mentions]
+        if len(common) != 0:
```
```suggestion
        if mentioned_afk_users:
```

Rapptz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,116 @@
+        common = [afk_user for afk_user in self.afk.keys()
+                  if afk_user in mentions]
+        if len(common) != 0:
+            akf_users = ",".join(str(self.bot.get_user(c)) for c in common)
```
This has a typo. Luckily, with `mentions` now being a dictionary you can get the resolved object without resorting to cache.

Rapptz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,116 @@
+                  if afk_user in mentions]
+        if len(common) != 0:
+            akf_users = ",".join(str(self.bot.get_user(c)) for c in common)
+            await message.channel.send(f"{akf_users} {'is' if len(common) == 1 else 'are'} afk.")
```
No f-strings.

Also this line is doing too much.

Daggy1234 commented:
A few comments:
- Now with the 3.8 confirmed, I stuck to f strings
- Fixed the comments as mentioned
- In the case of that weird Inbuilt error handling, I just added a note about which error would be raised, and assumed they would refer to the error handling example. I can add a cog_specific error handling example, but felt that would be out of scope. 


As for one of your comments, it would be very nice to have an event that dispatches when there is a message that isn't a command! Perhaps `on_message_without_command`.

Gobot1234 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,115 @@
+
+        # If an afk user is mentioned in a message, let the author of the message know they are afk
+        mentions = {user.id: user for user in message.mentions}
+        mentioned_afk_users = [user_id for user_id in self.afk_reasons.keys() if user_id in mentions]
```
```suggestion
        mentioned_afk_users = [user_id for user_id in self.afk_reasons if user_id in mentions]
```
.keys() isn't necessary here

NCPlayz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,115 @@
+        # Look at the error handling example
+        mentions = discord.AllowedMentions.none()
+        try:
+            afk = self.afk_reasons[user.id]
+            return await ctx.send(f"{user.name} is afk with reason `{afk}`", allowed_mentions=mentions)
```
```suggestion
            reason = self.afk_reasons[user.id]
            return await ctx.send(f"{user.name} is afk with reason `{reason}`", allowed_mentions=mentions)
```

NCPlayz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,115 @@
+            return
+
+        # If an afk user has sent a message, remove their afk
+        message_id = message.author.id
+        afk_status = self.afk_reasons.get(message_id)
+        if afk_status:
+            await message.channel.send(f"Welcome back {message.author.mention}! I have removed your afk status.")
+            self.afk_reasons.pop(message_id)
```
```suggestion
        user_id = message.author.id
        afk_status = self.afk_reasons.get(user_id)
        if afk_status:
            await message.channel.send(f"Welcome back {message.author.mention}! I have removed your afk status.")
            self.afk_reasons.pop(user_id)
```

Gobot1234 replied:
```suggestion
        if message_id in self.afk_reasons:
            await message.channel.send(f"Welcome back {message.author.mention}! I have removed your afk status.")
            del self.afk_reasons[message_id]
```

NCPlayz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,115 @@
+            return
+
+        # If an afk user is mentioned in a message, let the author of the message know they are afk
+        mentions = {user.id: user for user in message.mentions}
+        mentioned_afk_users = [user_id for user_id in self.afk_reasons.keys() if user_id in mentions]
+        if mentioned_afk_users:
+            plural_or_singular = "is" if len(mentioned_afk_users) == 1 else "are"
+            akf_users = ",".join(str(self.bot.get_user(c)) for c in mentioned_afk_users)
+            await message.channel.send(f"{akf_users} {plural_or_singular} afk.")
```
```suggestion
        mentioned_afk_users = [user for user in message.mentions if user.id in self.afk_reasons]
        if mentioned_afk_users:
            plural_or_singular = "is" if len(mentioned_afk_users) == 1 else "are"
            # formats each user as Username#0000
            formatted_users = [str(user) for user in mentioned_afk_users]
            afk_users = ",".join(formatted_users)
            await message.channel.send(f"{afk_users} {plural_or_singular} afk.")
```

This seems better than needing to `get_user` when we already have user object references from the mentions.

NCPlayz reviewed (changes requested):
Generally, discord.py uses apostrophes over quotation marks for strings.

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,115 @@
+
+# Cogs have to be explicitly added.
+bot.add_cog(Miscellaneous(bot))
+bot.run('token')
```
```suggestion
bot.run("token")
```
Consistency.

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,115 @@
+            afk = self.afk_reasons[user.id]
+            return await ctx.send(f"{user.name} is afk with reason `{afk}`", allowed_mentions=mentions)
+        except KeyError:
+            return await ctx.send(f'{user.name} is currently not afk', allowed_mentions=mentions)
```
```suggestion
            return await ctx.send(f"{user.name} is currently not afk", allowed_mentions=mentions)
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,115 @@
+from discord.ext import commands
+import random
+
+description = '''An example bot to showcase the discord.ext.commands extension
```
```suggestion
description = """An example bot to showcase the discord.ext.commands extension
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,115 @@
+module.
+There are a number of utility commands being showcased here.
+The example uses cogs to organize the code
+'''
```
```suggestion
"""
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,115 @@
+    # They must also pass a `self` parameter since it's within a class.
+    @commands.command()
+    async def hello(self, ctx):
+        """Adds two numbers together"""
```
```suggestion
        """Sends a hello response."""
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,115 @@
+    async def _user(self, ctx, *, user: discord.User):
+        """Check a user's afk status"""
+        # If a user isn't found an error will be thrown Handle that in an error handler.
+        # If the user input isn't present `MissingRequiredArgument` will be thrown
```
```suggestion
        # If the user input isn't present `MissingRequiredArgument` will be thrown.
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,115 @@
+        # If a user isn't found an error will be thrown Handle that in an error handler.
+        # If the user input isn't present `MissingRequiredArgument` will be thrown
+        # If a value that cannot be parsed to a member is present, `UserNotFound` will be thrown.
+        # Look at the error handling example
```
```suggestion
        # See the error handling example for how it will be handled.
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,115 @@
+    # Note that `self` has to be passed for these as well.
+    @commands.Cog.listener()
+    async def on_message(self, message):
+        # Don't want to reply to bots
```
```suggestion
        # Don't want to reply to bots:
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,115 @@
+        # Don't want to reply to bots
+        if message.author.bot:
+            return
+        # Ensure that the message wasn't the afk command
```
```suggestion
        # Ensure that the message wasn't the afk command:
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,115 @@
+        if ctx.valid:
+            return
+
+        # If an afk user has sent a message, remove their afk
```
```suggestion
        # If an afk user has sent a message, remove their afk:
```

mesub7 reviewed (changes requested):
Better, but a few single quotes  (amongst other nits) still remain.

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,113 @@
+intents = discord.Intents.default()
+intents.members = True
+
+bot = commands.Bot(command_prefix="?", description=description, intents=intents)
```
```suggestion
bot = commands.Bot(command_prefix='?', description=description, intents=intents)
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,113 @@
+    @commands.command()
+    async def hello(self, ctx):
+        """Have the bot greet you."""
+        await ctx.send("Hello am I an example bot!")
```
```suggestion
        await ctx.send('Hello am I an example bot!')
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,113 @@
+        if user_id not in self.afk_reasons:
+            self.afk_reasons[user_id] = reason
+            mentions = discord.AllowedMentions.none()
+            await ctx.send(f"I've set your afk status to `{reason}`", allowed_mentions=mentions)
```
```suggestion
            await ctx.send(f'I\'ve set your afk status to `{reason}`', allowed_mentions=mentions)
```

Gobot1234 replied:
This change isn't necessary

Daggy1234 replied:
Might as well for consistency, and single quote useage

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,113 @@
+            mentions = discord.AllowedMentions.none()
+            await ctx.send(f"I've set your afk status to `{reason}`", allowed_mentions=mentions)
+        else:
+            await ctx.send("You are already afk.")
```
```suggestion
            await ctx.send('You are already afk.')
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,113 @@
+        else:
+            await ctx.send("You are already afk.")
+
+    # Subcommand for the group
```
```suggestion
    # Subcommand for the group:
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,113 @@
+            await ctx.send("You are already afk.")
+
+    # Subcommand for the group
+    @afk.command(name="user")
```
```suggestion
    @afk.command(name='user')
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,113 @@
+        mentions = discord.AllowedMentions.none()
+        try:
+            reason = self.afk_reasons[user.id]
+            return await ctx.send(f"{user.name} is afk with reason `{reason}`", allowed_mentions=mentions)
```
```suggestion
            return await ctx.send(f'{user.name} is afk with reason `{reason}`', allowed_mentions=mentions)
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,113 @@
+            reason = self.afk_reasons[user.id]
+            return await ctx.send(f"{user.name} is afk with reason `{reason}`", allowed_mentions=mentions)
+        except KeyError:
+            return await ctx.send(f"{user.name} is currently not afk", allowed_mentions=mentions)
```
```suggestion
            return await ctx.send(f'{user.name} is currently not afk', allowed_mentions=mentions)
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,113 @@
+        # If an afk user has sent a message, remove their afk:
+        user_id = message.author.id
+        if user_id in self.afk_reasons:
+            await message.channel.send(f"Welcome back {message.author.mention}! I have removed your afk status.")
```
```suggestion
            await message.channel.send(f'Welcome back {message.author.mention}! I have removed your afk status.')
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,113 @@
+
+# Cogs have to be explicitly added.
+bot.add_cog(Miscellaneous(bot))
+bot.run("token")
```
```suggestion
bot.run('token')
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,113 @@
+        mentions = {user.id: user for user in message.mentions}
+        mentioned_afk_users = [user_id for user_id in self.afk_reasons if user_id in mentions]
+        if mentioned_afk_users:
+            plural_or_singular = "is" if len(mentioned_afk_users) == 1 else "are"
```
```suggestion
            plural_or_singular = 'is' if len(mentioned_afk_users) == 1 else 'are'
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,113 @@
+            plural_or_singular = "is" if len(mentioned_afk_users) == 1 else "are"
+            # formats each user as Username#0000
+            formatted_users = [str(bot.get_user(user)) for user in mentioned_afk_users]
+            afk_users = ",".join(formatted_users)
```
```suggestion
            afk_users = ','.join(formatted_users)
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,113 @@
+            # formats each user as Username#0000
+            formatted_users = [str(bot.get_user(user)) for user in mentioned_afk_users]
+            afk_users = ",".join(formatted_users)
+            await message.channel.send(f"{afk_users} {plural_or_singular} afk.")
```
```suggestion
            await message.channel.send(f'{afk_users} {plural_or_singular} afk.')
```

mesub7 commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,113 @@
+        # Don"t want to reply to bots
+        if message.author.bot:
+            return
+        # Ensure that the message wasn"t the afk command:
```
```suggestion
        # Ensure that the message wasn't the afk command:
```

mesub7 reviewed (changes requested):
My apologies, it seems like all double strings **should** be single strings with the exception of doc(triple) strings, in line with the other examples in the repo (as well as how it was before you re-drafted the proposed example). I've gone ahead and made the changes.

Daggy1234 commented:
Small necrobump here, anything I should add?

scrazzz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,113 @@
+    @commands.command()
+    async def hello(self, ctx):
+        """Have the bot greet you."""
+        await ctx.send('Hello am I an example bot!')
```
```suggestion
        await ctx.send('Hello I\'m an example bot!')
```

NCPlayz replied:
The better change would have been without the apostrophe.

scrazzz commented on examples/basic_cog.py:
```diff
@@ -0,0 +1,113 @@
+            plural_or_singular = 'is' if len(mentioned_afk_users) == 1 else 'are'
+            # formats each user as Username#0000
+            formatted_users = [str(bot.get_user(user)) for user in mentioned_afk_users]
+            afk_users = ','.join(formatted_users)
```
```suggestion
            afk_users = ', '.join(formatted_users)
```
